File: repo/frontend/src/api/operations.ts
import { apiGet, apiPatch, apiPost } from "./client";

export type RetentionPolicy = {
  id: string;
  org_id: string;
  itinerary_retention_days: number;
  updated_by_user_id: string | null;
  created_at: string;
  updated_at: string;
};

export type RetentionRun = {
  id: string;
  org_id: string;
  initiated_by_user_id: string | null;
  status: string;
  deleted_itinerary_count: number;
  summary: string | null;
  started_at: string;
  completed_at: string | null;
};

export type BackupRun = {
  id: string;
  org_id: string;
  initiated_by_user_id: string | null;
  trigger_kind: string;
  status: string;
  backup_file_name: string | null;
  backup_file_path: string | null;
  encrypted_size_bytes: number | null;
  rotated_file_count: number;
  summary: string | null;
  started_at: string;
  completed_at: string | null;
};

export type RestoreRun = {
  id: string;
  org_id: string;
  initiated_by_user_id: string | null;
  status: string;
  backup_file_name: string;
  restored_table_count: number;
  summary: string | null;
  started_at: string;
  completed_at: string | null;
};

export type AuditEvent = {
  id: string;
  org_id: string;
  project_id: string | null;
  actor_user_id: string | null;
  action_type: string;
  resource_type: string;
  resource_id: string | null;
  request_method: string;
  request_path: string;
  status_code: number;
  detail_summary: string | null;
  metadata_json: Record<string, unknown> | null;
  occurred_at: string;
};

export type LineageEvent = {
  id: string;
  org_id: string;
  project_id: string | null;
  dataset_id: string | null;
  itinerary_id: string | null;
  created_by_user_id: string | null;
  event_type: string;
  entity_type: string;
  entity_id: string | null;
  payload: Record<string, unknown>;
  occurred_at: string;
};

export const operationsApi = {
  getRetentionPolicy: () => apiGet<RetentionPolicy>("/api/ops/retention-policy"),
  updateRetentionPolicy: (itineraryRetentionDays: number) =>
    apiPatch<RetentionPolicy>("/api/ops/retention-policy", { itinerary_retention_days: itineraryRetentionDays }),
  runRetentionNow: () => apiPost<RetentionRun>("/api/ops/retention/run", {}),
  listRetentionRuns: (limit = 20) => apiGet<RetentionRun[]>(`/api/ops/retention/runs?limit=${encodeURIComponent(String(limit))}`),

  runBackupNow: () => apiPost<BackupRun>("/api/ops/backups/run", {}),
  listBackupRuns: (limit = 20) => apiGet<BackupRun[]>(`/api/ops/backups/runs?limit=${encodeURIComponent(String(limit))}`),

  runRestore: (backupFileName: string) => apiPost<RestoreRun>("/api/ops/restore", { backup_file_name: backupFileName }),
  listRestoreRuns: (limit = 20) => apiGet<RestoreRun[]>(`/api/ops/restore/runs?limit=${encodeURIComponent(String(limit))}`),

  listAuditEvents: (limit = 100) => apiGet<AuditEvent[]>(`/api/ops/audit/events?limit=${encodeURIComponent(String(limit))}`),
  listLineageEvents: (limit = 100) =>
    apiGet<LineageEvent[]>(`/api/ops/lineage/events?limit=${encodeURIComponent(String(limit))}`)
};

File: repo/backend/app/schemas/operations.py
from datetime import datetime

from pydantic import BaseModel, Field


class AuditEventOut(BaseModel):
    id: str
    org_id: str
    project_id: str | None
    actor_user_id: str | None
    action_type: str
    resource_type: str
    resource_id: str | None
    request_method: str
    request_path: str
    status_code: int
    detail_summary: str | None
    metadata_json: dict | None
    occurred_at: datetime


class LineageEventOut(BaseModel):
    id: str
    org_id: str
    project_id: str | None
    dataset_id: str | None
    itinerary_id: str | None
    created_by_user_id: str | None
    event_type: str
    entity_type: str
    entity_id: str | None
    payload: dict
    occurred_at: datetime


class RetentionPolicyOut(BaseModel):
    id: str
    org_id: str
    itinerary_retention_days: int
    audit_retention_days: int
    lineage_retention_days: int
    updated_by_user_id: str | None
    created_at: datetime
    updated_at: datetime


class RetentionPolicyUpdateRequest(BaseModel):
    itinerary_retention_days: int = Field(ge=30, le=3650)


class RetentionRunOut(BaseModel):
    id: str
    org_id: str
    initiated_by_user_id: str | None
    status: str
    deleted_itinerary_count: int
    deleted_audit_event_count: int
    deleted_lineage_event_count: int
    summary: str | None
    started_at: datetime
    completed_at: datetime | None


class BackupRunOut(BaseModel):
    id: str
    org_id: str
    initiated_by_user_id: str | None
    trigger_kind: str
    status: str
    backup_file_name: str | None
    backup_file_path: str | None
    encrypted_size_bytes: int | None
    rotated_file_count: int
    summary: str | None
    started_at: datetime
    completed_at: datetime | None


class RestoreRunOut(BaseModel):
    id: str
    org_id: str
    initiated_by_user_id: str | None
    status: str
    backup_file_name: str
    restored_table_count: int
    summary: str | None
    started_at: datetime
    completed_at: datetime | None


class RestoreRequest(BaseModel):
    backup_file_name: str = Field(min_length=1, max_length=255)
